File: client-go/src/main/client/utils/field.go
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package utils

import (
	"bytes"
	"encoding/binary"
	log "github.com/sirupsen/logrus"
)

type Field struct {
	DataType string
}

var (
	boolV   bool
	intV    int32
	longV   int64
	floatV  float32
	doubleV float64
	binaryV []byte
)

func Copy(field Field) Field {
	outField := NewField(field.DataType)
	if outField.DataType != "" {
		dateType := outField.DataType
		switch dateType {
		case "BOOLEAN":
			outField.SetBoolV(field.GetBoolV())
			break
		case "INT32":
			outField.SetIntV(field.GetIntV())
			break
		case "INT64":
			outField.SetLongV(field.GetLongV())
			break
		case "FLOAT":
			outField.SetFloatV(field.GetFloatV())
			break
		case "DOUBLE":
			outField.SetDoubleV(field.GetDoubleV())
			break
		case "TEXT":
			outField.SetBinaryV(field.GetBinaryV())
			break
		default:
			log.Error("not support dataType")
		}
	}
	return outField
}

func NewField(DataType string) Field {
	return Field{DataType}
}

func (f *Field) GetBoolV() bool {
	return boolV
}

func (f *Field) SetBoolV(boolVal bool) {
	boolV = boolVal
}

func (f *Field) GetBinaryV() []byte {
	return binaryV
}

func (f *Field) SetBinaryV(binaryVal []byte) {
	binaryV = binaryVal
}

func (f *Field) GetFloatV() float32 {
	return floatV
}

func (f *Field) SetFloatV(floatVal float32) {
	floatV = floatVal
}

func (f *Field) GetDoubleV() float64 {
	return doubleV
}

func (f *Field) SetDoubleV(doubleVal float64) {
	doubleV = doubleVal
}

func (f *Field) GetIntV() int32 {
	return intV
}

func (f *Field) SetIntV(intVal int32) {
	intV = intVal
}

func (f *Field) GetLongV() int64 {
	return longV
}

func (f *Field) SetLongV(longVal int64) {
	longV = longVal
}

func (f *Field) ToString() string {
	return f.GetStringValue()
}

func (f *Field) GetStringValue() string {
	if f.DataType == "" {
		return ""
	}
	dateType := f.DataType
	buf := bytes.NewBuffer([]byte{})
	switch dateType {
	case "BOOLEAN":
		binary.Write(buf, binary.BigEndian, &boolV)
		break
	case "INT32":
		binary.Write(buf, binary.BigEndian, &intV)
		break
	case "INT64":
		binary.Write(buf, binary.BigEndian, &longV)
		break
	case "FLOAT":
		binary.Write(buf, binary.BigEndian, &floatV)
		break
	case "DOUBLE":
		binary.Write(buf, binary.BigEndian, &doubleV)
		break
	case "TEXT":
		buf.Write(binaryV)
		break
	default:
		log.Error("not support dataType")
	}
	return buf.String()
}

func (f *Field) GetField(value interface{}, dateType string) *Field {
	if value == nil {
		return nil
	}
	field := NewField(dateType)
	switch dateType {
	case "BOOLEAN":
		field.SetBoolV(value.(bool))
		break
	case "INT32":
		field.SetIntV(int32(value.(int)))
		break
	case "INT64":
		field.SetLongV(int64(value.(int)))
		break
	case "FLOAT":
		field.SetFloatV(float32(value.(float64)))
		break
	case "DOUBLE":
		field.SetDoubleV(value.(float64))
		break
	case "TEXT":
		field.SetBinaryV([]byte(value.(string)))
		break
	default:
		log.Error("not support dataType")
	}
	return &field
}

func (f *Field) GetObjectValue(dateType string) interface{} {
	if f.DataType == "" {
		return nil
	}
	switch dateType {
	case "BOOLEAN":
		return f.GetBoolV()
		break
	case "INT32":
		return f.GetIntV()
		break
	case "INT64":
		return f.GetLongV()
		break
	case "FLOAT":
		return f.GetFloatV()
		break
	case "DOUBLE":
		return f.GetDoubleV()
		break
	case "TEXT":
		return f.GetBinaryV()
		break
	default:
		log.Error("not support dataType")
		return nil
	}
	return nil
}

func (f *Field) IsNull() bool {
	return f.DataType == ""
}

File: iotdb-client/client-py/iotdb/utils/Field.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

# for package
from iotdb.utils.IoTDBConstants import TSDataType
from iotdb.tsfile.utils.DateUtils import parse_int_to_date
import numpy as np
import pandas as pd


class Field(object):
    def __init__(self, data_type, value=None):
        """
        :param data_type: TSDataType
        """
        self.__data_type = data_type
        self.value = value

    @staticmethod
    def copy(field):
        output = Field(field.get_data_type())
        if output.get_data_type() is not None:
            if output.get_data_type() == TSDataType.BOOLEAN:
                output.set_bool_value(field.get_bool_value())
            elif (
                output.get_data_type() == TSDataType.INT32
                or output.get_data_type() == TSDataType.DATE
            ):
                output.set_int_value(field.get_int_value())
            elif (
                output.get_data_type() == TSDataType.INT64
                or output.get_data_type() == TSDataType.TIMESTAMP
            ):
                output.set_long_value(field.get_long_value())
            elif output.get_data_type() == TSDataType.FLOAT:
                output.set_float_value(field.get_float_value())
            elif output.get_data_type() == TSDataType.DOUBLE:
                output.set_double_value(field.get_double_value())
            elif (
                output.get_data_type() == TSDataType.TEXT
                or output.get_data_type() == TSDataType.STRING
                or output.get_data_type() == TSDataType.BLOB
            ):
                output.set_binary_value(field.get_binary_value())
            else:
                raise Exception(
                    "unsupported data type {}".format(output.get_data_type())
                )
        return output

    def get_data_type(self):
        return self.__data_type

    def is_null(self):
        return self.__data_type is None or self.value is None or self.value is pd.NA

    def set_bool_value(self, value: bool):
        self.value = value

    def get_bool_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.BOOLEAN
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return self.value

    def set_int_value(self, value: int):
        self.value = value

    def get_int_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.INT32
            and self.__data_type != TSDataType.DATE
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return np.int32(self.value)

    def set_long_value(self, value: int):
        self.value = value

    def get_long_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.INT64
            and self.__data_type != TSDataType.TIMESTAMP
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return np.int64(self.value)

    def set_float_value(self, value: float):
        self.value = value

    def get_float_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.FLOAT
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return np.float32(self.value)

    def set_double_value(self, value: float):
        self.value = value

    def get_double_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.DOUBLE
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return np.float64(self.value)

    def set_binary_value(self, value: bytes):
        self.value = value

    def get_binary_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.TEXT
            and self.__data_type != TSDataType.STRING
            and self.__data_type != TSDataType.BLOB
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return self.value

    def get_date_value(self):
        if self.__data_type is None:
            raise Exception("Null Field Exception!")
        if (
            self.__data_type != TSDataType.DATE
            or self.value is None
            or self.value is pd.NA
        ):
            return None
        return parse_int_to_date(self.value)

    def get_string_value(self):
        if self.__data_type is None or self.value is None or self.value is pd.NA:
            return "None"
        # TEXT, STRING
        elif self.__data_type == 5 or self.__data_type == 11:
            return self.value.decode("utf-8")
        # BLOB
        elif self.__data_type == 10:
            return str(hex(int.from_bytes(self.value, byteorder="big")))
        else:
            return str(self.get_object_value(self.__data_type))

    def __str__(self):
        return self.get_string_value()

    def get_object_value(self, data_type):
        """
        :param data_type: TSDataType
        """
        if self.__data_type is None or self.value is None or self.value is pd.NA:
            return None
        if data_type == 0:
            return bool(self.value)
        elif data_type == 1:
            return np.int32(self.value)
        elif data_type == 2 or data_type == 8:
            return np.int64(self.value)
        elif data_type == 3:
            return np.float32(self.value)
        elif data_type == 4:
            return np.float64(self.value)
        elif data_type == 9:
            return parse_int_to_date(self.value)
        elif data_type == 5 or data_type == 10 or data_type == 11:
            return self.value
        else:
            raise RuntimeError("Unsupported data type:" + str(data_type))

    @staticmethod
    def get_field(value, data_type):
        """
        :param value: field value corresponding to the data type
        :param data_type: TSDataType
        """
        if value is None or value is pd.NA:
            return None
        field = Field(data_type, value)
        return field
